# Домашнее задание к лекции "Базовые понятия статистики"
Будем осуществлять работу с непростым [набором данных](https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.csv) о состоянии здоровья лошадей, испытывающих кишечные колики. Цель – максимально корректно заполнить пропуски.

### Задание 1. Загрузка данных

Изучить представленный набор данных на основе [описания его столбцов](https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.names), загрузить его и оставить 8 столбцов для дальнейшего изучения: `surgery?`, `Age`, `rectal temperature`, `pulse`, `respiratory rate`, `temperature of extremities`, `pain`, `outcome`. 

### Задание 2. Первичное изучение данных

Проанализировать значения по столбцам, рассчитать базовые статистики, найти выбросы.


### Задание 3. Работа с пропусками

Рассчитать количество пропусков для всех выбранных столбцов. Принять и обосновать решение о методе заполнения пропусков по каждому столбцу на основе рассчитанных статистик и возможной взаимосвязи значений в них. Сформировать датафрейм, в котором пропуски будут отсутствовать.


In [417]:
import pandas as pd
header_list = ['surgery?','Age','Hospital Number','rectal temperature','pulse','respiratory rate','temperature of extremities','peripheral pulse','mucous membranes','capillary refill time','pain','peristalsis','abdominal distension','nasogastric tube','nasogastric reflux','nasogastric reflux PH','rectal examination - feces','abdomen','packed cell volume','total protein','abdominocentesis appearance','abdomcentesis total protein','outcome','surgical lesion?','type of lesion1','type of lesion2','type of lesion3','cp_data']
horses_df = pd.read_csv('horse_data.csv', names=header_list,na_values='?' )
horses_df.head(10)

surgery?  Age  Hospital Number  rectal temperature  pulse  \
0       2.0    1           530101                38.5   66.0   
1       1.0    1           534817                39.2   88.0   
2       2.0    1           530334                38.3   40.0   
3       1.0    9          5290409                39.1  164.0   
4       2.0    1           530255                37.3  104.0   
5       2.0    1           528355                 NaN    NaN   
6       1.0    1           526802                37.9   48.0   
7       1.0    1           529607                 NaN   60.0   
8       2.0    1           530051                 NaN   80.0   
9       2.0    9          5299629                38.3   90.0   

   respiratory rate  temperature of extremities  peripheral pulse  \
0              28.0                         3.0               3.0   
1              20.0                         NaN               NaN   
2              24.0                         1.0               1.0   
3              84.0                         4.0               1.0   
4              35.0                         NaN               NaN   
5               NaN                         2.0               1.0   
6              16.0                         1.0               1.0   
7               NaN                         3.0               NaN   
8              36.0                         3.0               4.0   
9               NaN                         1.0               NaN   

   mucous membranes  capillary refill time  ...  packed cell volume  \
0               NaN                    2.0  ...                45.0   
1               4.0                    1.0  ...                50.0   
2               3.0                    1.0  ...                33.0   
3               6.0                    2.0  ...                48.0   
4               6.0                    2.0  ...                74.0   
5               3.0                    1.0  ...                 NaN   
6               1.0                    1.0  ...                37.0   
7               NaN                    1.0  ...                44.0   
8               3.0                    1.0  ...                38.0   
9               1.0                    1.0  ...                40.0   

   total protein  abdominocentesis appearance  abdomcentesis total protein  \
0            8.4                          NaN                          NaN   
1           85.0                          2.0                          2.0   
2            6.7                          NaN                          NaN   
3            7.2                          3.0                          5.3   
4            7.4                          NaN                          NaN   
5            NaN                          NaN                          NaN   
6            7.0                          NaN                          NaN   
7            8.3                          NaN                          NaN   
8            6.2                          NaN                          NaN   
9            6.2                          1.0                          2.2   

   outcome  surgical lesion?  type of lesion1  type of lesion2  \
0      2.0                 2            11300                0   
1      3.0                 2             2208                0   
2      1.0                 2                0                0   
3      2.0                 1             2208                0   
4      2.0                 2             4300                0   
5      1.0                 2                0                0   
6      1.0                 1             3124                0   
7      2.0                 1             2208                0   
8      3.0                 1             3205                0   
9      1.0                 2                0                0   

   type of lesion3  cp_data  
0                0        2  
1                0        2  
2                0        1  
3                0        1  
4                0  

In [418]:
#в датафрейме оставляем только необходимые столбцы по заданию
drop_headers_list= ['Hospital Number','peripheral pulse','mucous membranes','capillary refill time','peristalsis','abdominal distension','nasogastric tube','nasogastric reflux','nasogastric reflux PH','rectal examination - feces','abdomen','packed cell volume','total protein','abdominocentesis appearance','abdomcentesis total protein','surgical lesion?','type of lesion1','type of lesion2','type of lesion3','cp_data']
horses_df.drop(drop_headers_list, axis=1,inplace=True)
horses_df.head()

surgery?  Age  rectal temperature  pulse  respiratory rate  \
0       2.0    1                38.5   66.0              28.0   
1       1.0    1                39.2   88.0              20.0   
2       2.0    1                38.3   40.0              24.0   
3       1.0    9                39.1  164.0              84.0   
4       2.0    1                37.3  104.0              35.0   

   temperature of extremities  pain  outcome  
0                         3.0   5.0      2.0  
1                         NaN   3.0      3.0  
2                         1.0   3.0      1.0  
3                         4.0   2.0      2.0  
4                         NaN   NaN      2.0

In [419]:
horses_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery?                    299 non-null    float64
 1   Age                         300 non-null    int64  
 2   rectal temperature          240 non-null    float64
 3   pulse                       276 non-null    float64
 4   respiratory rate            242 non-null    float64
 5   temperature of extremities  244 non-null    float64
 6   pain                        245 non-null    float64
 7   outcome                     299 non-null    float64
dtypes: float64(7), int64(1)
memory usage: 18.9 KB


In [420]:
horses_df.describe()

surgery?         Age  rectal temperature       pulse  \
count  299.000000  300.000000          240.000000  276.000000   
mean     1.397993    1.640000           38.167917   71.913043   
std      0.490305    2.173972            0.732289   28.630557   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.800000   48.000000   
50%      1.000000    1.000000           38.200000   64.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    9.000000           40.800000  184.000000   

       respiratory rate  temperature of extremities       pain     outcome  
count        242.000000                  244.000000  245.00000  299.000000  
mean          30.417355                    2.348361    2.95102    1.551839  
std           17.642231                    1.045054    1.30794    0.737187  
min            8.000000                    1.000000    1.00000    1.000000  
25%           18.500000                    1.000000    2.00000    1.000000  
50%           24.500000                    3.000000    3.00000    1.000000  
75%           36.000000                    3.000000    4.00000    2.000000  
max           96.000000                    4.000000    5.00000    3.000000

## Выбросы.

Воспользуемся описанием значений столбцов.

столбец Age. Максимальное значение 9, а по описанию 2. Выбросы есть

столбец pulse. Максимальное значение 184, а по описанию 20-40. Выбросы есть

столбец respiratory rate. Максимальное значение 96, а по описанию 8-10 (хотя в описании написано, что 
полезность столбца сомнительна, из-за больших отклонений)
Все остальные столбцы в содержат значения в допустимой норме, исходя из описания.


  1:  surgery?
          1 = Yes, it had surgery
          2 = It was treated without surgery

  2:  Age
          1 = Adult horse
          2 = Young (< 6 months)
  4:  rectal temperature
          - linear
          - in degrees celsius.
          - An elevated temp may occur due to infection.
          - temperature may be reduced when the animal is in late shock
          - normal temp is 37.8
          - this parameter will usually change as the problem progresses
               eg. may start out normal, then become elevated because of
                   the lesion, passing back through the normal range as the
                   horse goes into shock
  5:  pulse
          - linear
          - the heart rate in beats per minute
          - is a reflection of the heart condition: 30 -40 is normal for adults
          - rare to have a lower than normal rate although athletic horses
            may have a rate of 20-25
          - animals with painful lesions or suffering from circulatory shock
            may have an elevated heart rate
  6:  respiratory rate
          - linear
          - normal rate is 8 to 10
          - usefulness is doubtful due to the great fluctuations

  7:  temperature of extremities
          - a subjective indication of peripheral circulation
          - possible values:
               1 = Normal
               2 = Warm
               3 = Cool
               4 = Cold
          - cool to cold extremities indicate possible shock
          - hot extremities should correlate with an elevated rectal temp.
   11: pain - a subjective judgement of the horse's pain level
          - possible values:
               1 = alert, no pain
               2 = depressed
               3 = intermittent mild pain
               4 = intermittent severe pain
               5 = continuous severe pain
          - should NOT be treated as a ordered or discrete variable!
          - In general, the more painful, the more likely it is to require
            surgery
          - prior treatment of pain may mask the pain level to some extent
   23: outcome
          - what eventually happened to the horse?
          - possible values:
               1 = lived
               2 = died
               3 = was euthanized


In [421]:
#найдем выбросы по столбцу Age (из описания макимально допустимое значение =2)
horses_df.loc[horses_df['Age']>2]

surgery?  Age  rectal temperature  pulse  respiratory rate  \
3         1.0    9                39.1  164.0              84.0   
9         2.0    9                38.3   90.0               NaN   
13        2.0    9                38.0   92.0              28.0   
16        1.0    9                 NaN  128.0              36.0   
23        1.0    9                38.3  130.0              60.0   
39        1.0    9                39.2  146.0              96.0   
41        2.0    9                39.0  150.0              72.0   
55        1.0    9                38.6  160.0              20.0   
74        1.0    9                 NaN    NaN               NaN   
75        1.0    9                39.7  100.0               NaN   
82        1.0    9                38.1  100.0              80.0   
103       1.0    9                38.0  140.0              68.0   
109       1.0    9                38.4   84.0              40.0   
135       1.0    9                38.1  136.0              48.0   
191       2.0    9                 NaN  100.0              44.0   
212       1.0    9                38.9  120.0              30.0   
227       1.0    9                 NaN    NaN               NaN   
229       1.0    9                38.5  120.0              70.0   
244       1.0    9                38.2  124.0              88.0   
252       1.0    9                38.8  124.0              36.0   
255       1.0    9                38.8  184.0              84.0   
262       2.0    9                38.6   88.0              28.0   
275       1.0    9                38.8  150.0              50.0   
287       2.0    9                39.5   84.0              30.0   

     temperature of extremities  pain  outcome  
3                           4.0   2.0      2.0  
9                           1.0   5.0      1.0  
13                          1.0   1.0      2.0  
16                          3.0   4.0      2.0  
23                          NaN   2.0      1.0  
39                          NaN   NaN      2.0  
41                          NaN   NaN      1.0  
55                          3.0   3.0      2.0  
74                          NaN   NaN      2.0  
75                          3.0   2.0      3.0  
82                          3.0   3.0      1.0  
103                         1.0   3.0      1.0  
109                         3.0   3.0      2.0  
135                         3.0   5.0      2.0  
191                         2.0   4.0      1.0  
212                         1.0   3.0      1.0  
227                         NaN   NaN      1.0  
229                         NaN   NaN      1.0  
244                         1.0   2.0      1.0  
252                         3.0   2.0      2.0  
255                         1.0   4.0      2.0  
262                         NaN   NaN      1.0  
275                         1.0   5.0      2.0  
287                         NaN   NaN      1.0

In [422]:
#посчитаем их
horses_df['Age'].value_counts()

1    276
9     24
Name: Age, dtype: int64

### принимаем решение выбросы по столбцу Age со значением '9' заменить на 1, 
т.к. это единственное значение,которое уже присутствует в столбце и соответствует лошади старше 6 месяцев

In [423]:
horses_df.loc[horses_df['Age'] == 9,'Age'] = 1
horses_df.describe()

surgery?    Age  rectal temperature       pulse  respiratory rate  \
count  299.000000  300.0          240.000000  276.000000        242.000000   
mean     1.397993    1.0           38.167917   71.913043         30.417355   
std      0.490305    0.0            0.732289   28.630557         17.642231   
min      1.000000    1.0           35.400000   30.000000          8.000000   
25%      1.000000    1.0           37.800000   48.000000         18.500000   
50%      1.000000    1.0           38.200000   64.000000         24.500000   
75%      2.000000    1.0           38.500000   88.000000         36.000000   
max      2.000000    1.0           40.800000  184.000000         96.000000   

       temperature of extremities       pain     outcome  
count                  244.000000  245.00000  299.000000  
mean                     2.348361    2.95102    1.551839  
std                      1.045054    1.30794    0.737187  
min                      1.000000    1.00000    1.000000  
25%                      1.000000    2.00000    1.000000  
50%                      3.000000    3.00000    1.000000  
75%                      3.000000    4.00000    2.000000  
max                      4.000000    5.00000    3.000000

In [424]:
# Посчитаем межквартильные размахи для определения выбросов по столбцу 'pulse'
q1 = horses_df['pulse'].quantile(0.25)
q3 = horses_df['pulse'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
remove_outliers = horses_df[horses_df['pulse'].between(lower_bound, upper_bound, inclusive=True)].sort_values('pulse')
remove_outliers

surgery?  Age  rectal temperature  pulse  respiratory rate  \
130       1.0    1                37.8   30.0              12.0   
232       1.0    1                38.5   30.0              18.0   
203       2.0    1                37.2   36.0               9.0   
276       1.0    1                38.0   36.0              16.0   
242       1.0    1                 NaN   36.0              32.0   
..        ...  ...                 ...    ...               ...   
148       1.0    1                38.3  132.0               NaN   
135       1.0    1                38.1  136.0              48.0   
44        1.0    1                35.4  140.0              24.0   
103       1.0    1                38.0  140.0              68.0   
39        1.0    1                39.2  146.0              96.0   

     temperature of extremities  pain  outcome  
130                         NaN   NaN      2.0  
232                         NaN   NaN      1.0  
203                         1.0   2.0      1.0  
276                         3.0   4.0      3.0  
242                         1.0   5.0      2.0  
..                          ...   ...      ...  
148                         NaN   2.0      1.0  
135                         3.0   5.0      2.0  
44                          3.0   4.0      3.0  
103                         1.0   3.0      1.0  
39                          NaN   NaN      2.0  

[271 rows x 8 columns]

In [425]:
#Что это за выбросы по столбцу 'pulse'?
pd.concat([horses_df, remove_outliers]).drop_duplicates(keep=False)

surgery?  Age  rectal temperature  pulse  respiratory rate  \
3         1.0    1                39.1  164.0              84.0   
5         2.0    1                 NaN    NaN               NaN   
41        2.0    1                39.0  150.0              72.0   
52        2.0    1                 NaN    NaN               NaN   
55        1.0    1                38.6  160.0              20.0   
58        1.0    1                 NaN    NaN              20.0   
93        2.0    1                 NaN    NaN               NaN   
115       2.0    1                 NaN    NaN              40.0   
117       2.0    1                39.5    NaN               NaN   
126       1.0    1                38.7    NaN               NaN   
151       1.0    1                39.3    NaN               NaN   
159       2.0    1                 NaN    NaN              12.0   
160       1.0    1                 NaN    NaN               NaN   
174       2.0    1                 NaN    NaN               NaN   
204       1.0    1                39.2    NaN              23.0   
216       2.0    1                 NaN    NaN               NaN   
255       1.0    1                38.8  184.0              84.0   
275       1.0    1                38.8  150.0              50.0   

     temperature of extremities  pain  outcome  
3                           4.0   2.0      2.0  
5                           2.0   2.0      1.0  
41                          NaN   NaN      1.0  
52                          1.0   1.0      1.0  
55                          3.0   3.0      2.0  
58                          4.0   5.0      2.0  
93                          3.0   5.0      2.0  
115                         3.0   3.0      1.0  
117                         3.0   3.0      3.0  
126                         3.0   5.0      1.0  
151                         4.0   4.0      2.0  
159                         1.0   2.0      1.0  
160                         NaN   4.0      1.0  
174                         3.0   4.0      2.0  
204                         3.0   4.0      1.0  
216                         NaN   NaN      1.0  
255                         1.0   4.0      2.0  
275                         1.0   5.0      2.0

### Выбросы по столбцу pulse оставляем, не корректируем 
    Интернет говорит,что высокий пульс у лошади  может до 180-240 ударов в минуту

In [426]:
# Посчитаем межквартильные размахи для определения выбросов по столбцу'respiratory rate'
q1 = horses_df['respiratory rate'].quantile(0.25)
q3 = horses_df['respiratory rate'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
remove_outliers = horses_df[horses_df['respiratory rate'].between(lower_bound, upper_bound, inclusive=True)].sort_values('respiratory rate')
remove_outliers

surgery?  Age  rectal temperature  pulse  respiratory rate  \
190       1.0    1                37.1   40.0               8.0   
203       2.0    1                37.2   36.0               9.0   
140       1.0    1                38.1   44.0               9.0   
89        2.0    1                38.3   42.0              10.0   
162       2.0    1                39.5   60.0              10.0   
..        ...  ...                 ...    ...               ...   
220       1.0    1                39.2   88.0              58.0   
20        1.0    1                39.9   72.0              60.0   
23        1.0    1                38.3  130.0              60.0   
185       1.0    1                38.5  120.0              60.0   
248       1.0    1                38.0  108.0              60.0   

     temperature of extremities  pain  outcome  
190                         NaN   3.0      1.0  
203                         1.0   2.0      1.0  
140                         3.0   2.0      1.0  
89                          1.0   1.0      1.0  
162                         3.0   3.0      1.0  
..                          ...   ...      ...  
220                         4.0   5.0      3.0  
20                          1.0   5.0      1.0  
23                          NaN   2.0      1.0  
185                         4.0   NaN      1.0  
248                         2.0   4.0      1.0  

[225 rows x 8 columns]

In [376]:
#Что это за выбросы по столбцу 'respiratory rate'?
pd.concat([horses_df, remove_outliers]).drop_duplicates(keep=False)

surgery?  Age  rectal temperature  pulse  respiratory rate  \
3         1.0    1                39.1  164.0              84.0   
5         2.0    1                 NaN    NaN               NaN   
7         1.0    1                 NaN   60.0               NaN   
9         2.0    1                38.3   90.0               NaN   
29        2.0    1                37.7   48.0               NaN   
32        1.0    1                37.2   60.0               NaN   
37        1.0    1                37.8   72.0               NaN   
38        2.0    1                38.6   52.0               NaN   
39        1.0    1                39.2  146.0              96.0   
40        1.0    1                 NaN   88.0               NaN   
41        2.0    1                39.0  150.0              72.0   
43        1.0    1                 NaN  120.0               NaN   
45        2.0    1                 NaN  120.0               NaN   
50        2.0    1                38.6   46.0               NaN   
52        2.0    1                 NaN    NaN               NaN   
61        2.0    1                38.5   60.0               NaN   
66        2.0    1                37.9   72.0               NaN   
72        1.0    1                37.7   56.0               NaN   
75        1.0    1                39.7  100.0               NaN   
81        1.0    1                37.3   40.0               NaN   
82        1.0    1                38.1  100.0              80.0   
84        1.0    1                37.8   60.0              80.0   
93        2.0    1                 NaN    NaN               NaN   
103       1.0    1                38.0  140.0              68.0   
106       1.0    1                38.3   52.0              96.0   
112       1.0    1                 NaN   40.0               NaN   
117       2.0    1                39.5    NaN               NaN   
120       1.0    1                39.4   54.0              66.0   
125       1.0    1                38.0   42.0              68.0   
126       1.0    1                38.7    NaN               NaN   
129       1.0    1                38.4   49.0               NaN   
134       1.0    1                 NaN  100.0               NaN   
137       1.0    1                38.0   48.0               NaN   
138       2.0    1                38.0   56.0               NaN   
142       1.0    1                 NaN  120.0               NaN   
146       1.0    1                38.2   48.0               NaN   
148       1.0    1                38.3  132.0               NaN   
151       1.0    1                39.3    NaN               NaN   
152       1.0    1                 NaN  100.0               NaN   
160       1.0    1                 NaN    NaN               NaN   
164       1.0    1                38.7   76.0               NaN   
173       1.0    1                 NaN  104.0               NaN   
174       2.0    1                 NaN    NaN               NaN   
188       1.0    1                38.5   60.0               NaN   
201       1.0    1                38.5   86.0               NaN   
208       1.0    1                37.8   88.0              80.0   
216       2.0    1                 NaN    NaN               NaN   
229       1.0    1                38.5  120.0              70.0   
238       2.0    1                36.1   88.0               NaN   
244       1.0    1                38.2  124.0              88.0   
249       2.0    1                38.2   48.0               NaN   
254       1.0    1                 NaN   80.0               NaN   
255       1.0    1                38.8  184.0              84.0   
256       1.0    1                37.5   72.0               NaN   
265       2.0    1                 NaN   56.0              68.0   
271       1.0    1                37.7   80.0               NaN   
274       1.0    1                 NaN   76.0               NaN   
281       2.0    1                40.0   78.0               NaN   
284       2.0    1                38.5   54.0       

### Выбросы по столбцу пульс оставляем, не корректируем
интернет говорит,что частота дыхания у возбужденной лошади может быть повышенной.

## Пропуски

In [427]:
#посчитаем все пропуски
(horses_df.isna().mean() * 100).round(2)

surgery?                       0.33
Age                            0.00
rectal temperature            20.00
pulse                          8.00
respiratory rate              19.33
temperature of extremities    18.67
pain                          18.33
outcome                        0.33
dtype: float64

т.к у нас отсутствуют столбцы с пропусками более 80%, то все столбцы оставляем

In [469]:
#для категориальных столбцов temperature of extremities, pain, outcome, найдем модальные значения и затем решим
# о целесообразности замены отсутствующих значений на эти модальные значения
horses_df['temperature of extremities'].value_counts()

3.0    133
1.0     78
2.0     62
4.0     27
Name: temperature of extremities, dtype: int64

In [429]:
horses_df['pain'].value_counts()

3.0    67
2.0    59
5.0    42
4.0    39
1.0    38
Name: pain, dtype: int64

In [430]:
horses_df['outcome'].value_counts()


1.0    178
2.0     77
3.0     44
Name: outcome, dtype: int64

Кроме столбца outcome, нет ярко выраженных мод внутри столбцов
Будем искать взаимосвязь между различными столбцами- результат замены то этого только улучшится ;)

In [431]:
# Гипотеза- проверим взаимосвязь между столбцами pain и outcome 
print(horses_df.groupby(['pain'])['outcome'].median())

pain
1.0    1.0
2.0    1.0
3.0    1.0
4.0    2.0
5.0    2.0
Name: outcome, dtype: float64


можно сделать вывод, что при значениях pain(боль)-  4 (периодическая сильная боль) и 5(постоянная сильная боль), медианное значение столбца outcome равно 2 (смерть).
Поэтому недостающие значения в столбце outcome заполняеем исходя из значения столбца pain

In [432]:
#Сначала заполним недостающими значениями столбец outcome (у него самый маленький процент пропусков)
#найдем это значение 
horses_df.loc[((horses_df['pain'] == 4) | (horses_df['pain'] == 5)) & horses_df['outcome'].isna()]

surgery?  Age  rectal temperature  pulse  respiratory rate  \
132       NaN    1                38.0   48.0              20.0   

     temperature of extremities  pain  outcome  
132                         3.0   4.0      NaN

In [433]:
#Все действия будем производить с новым датафреймом, для быстрого отката в случае неправильного применения фильтра
#Заполним недостающие значениями столбца outcome
fill_horses_df = horses_df
fill_horses_df['outcome'].fillna(fill_horses_df.groupby(['pain'])['outcome'].transform('median'), inplace=True)
fill_horses_df.describe()

surgery?    Age  rectal temperature       pulse  respiratory rate  \
count  299.000000  300.0          240.000000  276.000000        242.000000   
mean     1.397993    1.0           38.167917   71.913043         30.417355   
std      0.490305    0.0            0.732289   28.630557         17.642231   
min      1.000000    1.0           35.400000   30.000000          8.000000   
25%      1.000000    1.0           37.800000   48.000000         18.500000   
50%      1.000000    1.0           38.200000   64.000000         24.500000   
75%      2.000000    1.0           38.500000   88.000000         36.000000   
max      2.000000    1.0           40.800000  184.000000         96.000000   

       temperature of extremities       pain     outcome  
count                  244.000000  245.00000  300.000000  
mean                     2.348361    2.95102    1.553333  
std                      1.045054    1.30794    0.736408  
min                      1.000000    1.00000    1.000000  
25%                      1.000000    2.00000    1.000000  
50%                      3.000000    3.00000    1.000000  
75%                      3.000000    4.00000    2.000000  
max                      4.000000    5.00000    3.000000

In [434]:
# Гипотеза- проверим взаимосвязь между столбцами surgery? и pain (тем более это описанов документации)
print(fill_horses_df.groupby(['surgery?'])['pain'].median())

surgery?
1.0    3.0
2.0    2.0
Name: pain, dtype: float64


Взаимосвязь есть- операция была (1) при высоком медианном значении боли(3), и не было(2) при низком значении боли(2)

In [435]:
#Заполним недостающие значениями столбца pain исходя из медианных значений столбца surgery
fill_horses_df['pain'].fillna(fill_horses_df.groupby(['surgery?'])['pain'].transform('median'), inplace=True)
fill_horses_df.describe()

surgery?    Age  rectal temperature       pulse  respiratory rate  \
count  299.000000  300.0          240.000000  276.000000        242.000000   
mean     1.397993    1.0           38.167917   71.913043         30.417355   
std      0.490305    0.0            0.732289   28.630557         17.642231   
min      1.000000    1.0           35.400000   30.000000          8.000000   
25%      1.000000    1.0           37.800000   48.000000         18.500000   
50%      1.000000    1.0           38.200000   64.000000         24.500000   
75%      2.000000    1.0           38.500000   88.000000         36.000000   
max      2.000000    1.0           40.800000  184.000000         96.000000   

       temperature of extremities        pain     outcome  
count                  244.000000  300.000000  300.000000  
mean                     2.348361    2.900000    1.553333  
std                      1.045054    1.203395    0.736408  
min                      1.000000    1.000000    1.000000  
25%                      1.000000    2.000000    1.000000  
50%                      3.000000    3.000000    1.000000  
75%                      3.000000    4.000000    2.000000  
max                      4.000000    5.000000    3.000000

In [436]:
# В подтверждение гипотезы роверим обратную взаимосвязь между столбцами pain(боль) и surgery(хтрургическое вмешательство)
print(horses_df.groupby(['pain'])['surgery?'].median())

pain
1.0    2.0
2.0    2.0
3.0    1.0
4.0    1.0
5.0    1.0
Name: surgery?, dtype: float64


взаимосвязь есть - что на уровнях боли 3 и выше всегда было хирургическое вмешательство. Восстановим отсутствующее значение в столбце 'surgery?' 

In [437]:
fill_horses_df['surgery?'].fillna(fill_horses_df.groupby(['pain'])['surgery?'].transform('median'), inplace=True)
fill_horses_df.describe()

surgery?    Age  rectal temperature       pulse  respiratory rate  \
count  300.000000  300.0          240.000000  276.000000        242.000000   
mean     1.396667    1.0           38.167917   71.913043         30.417355   
std      0.490023    0.0            0.732289   28.630557         17.642231   
min      1.000000    1.0           35.400000   30.000000          8.000000   
25%      1.000000    1.0           37.800000   48.000000         18.500000   
50%      1.000000    1.0           38.200000   64.000000         24.500000   
75%      2.000000    1.0           38.500000   88.000000         36.000000   
max      2.000000    1.0           40.800000  184.000000         96.000000   

       temperature of extremities        pain     outcome  
count                  244.000000  300.000000  300.000000  
mean                     2.348361    2.900000    1.553333  
std                      1.045054    1.203395    0.736408  
min                      1.000000    1.000000    1.000000  
25%                      1.000000    2.000000    1.000000  
50%                      3.000000    3.000000    1.000000  
75%                      3.000000    4.000000    2.000000  
max                      4.000000    5.000000    3.000000

In [ ]:
# 

In [438]:
# Гипотеза- проверим взаимосвязь между outcome  и temperature of extremities
print(horses_df.groupby(['outcome'])['temperature of extremities'].median())

outcome
1.0    2.0
2.0    3.0
3.0    3.0
Name: temperature of extremities, dtype: float64


Вывод- выявлена прямая связь-в случаях смертности (outcome =2 или 3),температура экскрементов (temperature of extremities) всегда была 3(это признак шока- из описания). В случае если outcome =1 (лошадь осталась живой), температура экскрементов =2 (норма)

In [439]:
#Восстановим отсутствующие значения в столбце 'temperature of extremities' 
fill_horses_df['temperature of extremities'].fillna(fill_horses_df.groupby(['outcome'])['temperature of extremities'].transform('median'), inplace=True)
fill_horses_df.describe()

surgery?    Age  rectal temperature       pulse  respiratory rate  \
count  300.000000  300.0          240.000000  276.000000        242.000000   
mean     1.396667    1.0           38.167917   71.913043         30.417355   
std      0.490023    0.0            0.732289   28.630557         17.642231   
min      1.000000    1.0           35.400000   30.000000          8.000000   
25%      1.000000    1.0           37.800000   48.000000         18.500000   
50%      1.000000    1.0           38.200000   64.000000         24.500000   
75%      2.000000    1.0           38.500000   88.000000         36.000000   
max      2.000000    1.0           40.800000  184.000000         96.000000   

       temperature of extremities        pain     outcome  
count                  300.000000  300.000000  300.000000  
mean                     2.363333    2.900000    1.553333  
std                      0.966663    1.203395    0.736408  
min                      1.000000    1.000000    1.000000  
25%                      1.000000    2.000000    1.000000  
50%                      3.000000    3.000000    1.000000  
75%                      3.000000    4.000000    2.000000  
max                      4.000000    5.000000    3.000000

In [440]:
# Гипотеза- проверим взаимосвязь между столбцами temperature of extremities и pulse
print(horses_df.groupby(['temperature of extremities'])['pulse'].median())

temperature of extremities
1.0    54.0
2.0    54.0
3.0    78.0
4.0    90.0
Name: pulse, dtype: float64


Вывод- выявлена прямая связь- чем ниже температура экскрементов(признак заболевания)- тем выше пульс

In [441]:
#Восстановим отсутствующие значения в столбце 'pulse' 
fill_horses_df['pulse'].fillna(fill_horses_df.groupby(['temperature of extremities'])['pulse'].transform('median'), inplace=True)
fill_horses_df.describe()


surgery?    Age  rectal temperature       pulse  respiratory rate  \
count  300.000000  300.0          240.000000  300.000000        242.000000   
mean     1.396667    1.0           38.167917   71.760000         30.417355   
std      0.490023    0.0            0.732289   27.701027         17.642231   
min      1.000000    1.0           35.400000   30.000000          8.000000   
25%      1.000000    1.0           37.800000   48.000000         18.500000   
50%      1.000000    1.0           38.200000   64.000000         24.500000   
75%      2.000000    1.0           38.500000   88.000000         36.000000   
max      2.000000    1.0           40.800000  184.000000         96.000000   

       temperature of extremities        pain     outcome  
count                  300.000000  300.000000  300.000000  
mean                     2.363333    2.900000    1.553333  
std                      0.966663    1.203395    0.736408  
min                      1.000000    1.000000    1.000000  
25%                      1.000000    2.000000    1.000000  
50%                      3.000000    3.000000    1.000000  
75%                      3.000000    4.000000    2.000000  
max                      4.000000    5.000000    3.000000

In [455]:
#Гипотеза - проверим утверждения из описании, что ректальная температура повышается во время инфекции,а во время болевого 
#шока(значение-pain) температура понижается
print(horses_df.groupby(['pain'])['rectal temperature'].median())

pain
1.0    38.2
2.0    38.3
3.0    38.2
4.0    38.0
5.0    38.0
Name: rectal temperature, dtype: float64


Вывод- гипотезу можно признать успешной. При высоких показателях боли медианная температура понижается.


In [456]:
#Восстановим отсутствующие значения в столбце 'rectal temperature' 
fill_horses_df['rectal temperature'].fillna(fill_horses_df.groupby(['pain'])['rectal temperature'].transform('median'), inplace=True)
fill_horses_df.describe()

surgery?    Age  rectal temperature       pulse  respiratory rate  \
count  300.000000  300.0          300.000000  300.000000        242.000000   
mean     1.396667    1.0           38.159333   71.760000         30.417355   
std      0.490023    0.0            0.657358   27.701027         17.642231   
min      1.000000    1.0           35.400000   30.000000          8.000000   
25%      1.000000    1.0           37.900000   48.000000         18.500000   
50%      1.000000    1.0           38.200000   64.000000         24.500000   
75%      2.000000    1.0           38.500000   88.000000         36.000000   
max      2.000000    1.0           40.800000  184.000000         96.000000   

       temperature of extremities        pain     outcome  
count                  300.000000  300.000000  300.000000  
mean                     2.363333    2.900000    1.553333  
std                      0.966663    1.203395    0.736408  
min                      1.000000    1.000000    1.000000  
25%                      1.000000    2.000000    1.000000  
50%                      3.000000    3.000000    1.000000  
75%                      3.000000    4.000000    2.000000  
max                      4.000000    5.000000    3.000000

Остался последний показатель- respiratory rate. В описании так и написано- полезность сомнительная из-за больших колебаний.
Поэтому заполним недостающие значения медианными.

In [467]:
fill_horses_df['respiratory rate'].fillna(fill_horses_df['respiratory rate'].median(), inplace=True)
fill_horses_df.describe()

surgery?    Age  rectal temperature       pulse  respiratory rate  \
count  300.000000  300.0          300.000000  300.000000        300.000000   
mean     1.396667    1.0           38.159333   71.760000         29.273333   
std      0.490023    0.0            0.657358   27.701027         16.010979   
min      1.000000    1.0           35.400000   30.000000          8.000000   
25%      1.000000    1.0           37.900000   48.000000         20.000000   
50%      1.000000    1.0           38.200000   64.000000         24.500000   
75%      2.000000    1.0           38.500000   88.000000         34.250000   
max      2.000000    1.0           40.800000  184.000000         96.000000   

       temperature of extremities        pain     outcome  
count                  300.000000  300.000000  300.000000  
mean                     2.363333    2.900000    1.553333  
std                      0.966663    1.203395    0.736408  
min                      1.000000    1.000000    1.000000  
25%                      1.000000    2.000000    1.000000  
50%                      3.000000    3.000000    1.000000  
75%                      3.000000    4.000000    2.000000  
max                      4.000000    5.000000    3.000000

Все, все значения заполнены.